crates/argmin/src/core/state/iterstate.rs

```diff
@@ -1356,9 +1359,11 @@ where
     /// # assert_eq!(state.counts, hm);
     /// ```
     fn func_counts<O>(&mut self, problem: &Problem<O>) {
-        for (k, &v) in problem.counts.iter() {
-            let count = self.counts.entry(k.to_string()).or_insert(0);
-            *count = v
+        if self.counting_enabled {
+            for (k, &v) in problem.counts.iter() {
+                let count = self.counts.entry(k.to_string()).or_insert(0);
```

Comment:
By the way, have you been able to find out whether hashing is the problem or the allocation of `.to_string()`? In case it was the latter, I was wondering if `Cow` would help?
If it's hashing, I was wondering if replacing `HashMap<String, u64>` with `HashMap<SomeEnum, u64>` would help, where 

```rust
enum SomeEnum {
    CostCount,
    GradientCount,
    OperatorCount,
    JacobianCount,
    HessianCount,
    AnnealCount,
    Other(String),
}
```

(or something along those lines). That would cover all the function calls that in argmin, but also allows one to count function calls defined in external code via `Other`. It also avoids `.to_string()`.

Reply:
As far as I remember it's 35% allocation and 65% hashing. `HashMap` is DoS resistant by default. Changing the hash function or switching to `BTreeMap` combined with enum approach you propose might help a bit I guess, but this changes public API.

Reply:
I don't mind changing the public API in that case. Why do you propose a `BTreeMap` instead of a `HashMap`? Hashing an enum should be quick (I expect it to use it's integer representation as a hash with a bit of additional overhead for hashing `Other(String)`). Anyway I think this goes beyond this PR and can certainly be left as a future improvement.